metricbeat/module/elasticsearch/cluster_stats/data_xpack.go

```diff
@@ -0,0 +1,339 @@
+				"master":            c.Int("master"),
+				"ingest":            c.Int("ingest"),
+			}),
+			"os": c.Dict("os", s.Schema{
+				"available_processors": c.Int("available_processors"),
+				"allocated_processors": c.Int("allocated_processors"),
+				"mem": c.Dict("mem", s.Schema{
+					"total_in_bytes": c.Int("total_in_bytes"),
+					"free_in_bytes":  c.Int("free_in_bytes"),
+					"used_in_bytes":  c.Int("used_in_bytes"),
+					"free_percent":   c.Int("free_percent"),
+					"used_percent":   c.Int("used_percent"),
+				}),
+			}),
+			"process": c.Dict("process", s.Schema{
+				"cpu": c.Dict("cpu", s.Schema{
+					"percent": c.Int("percent"),
+				}),
+				"open_file_descriptors": c.Dict("open_file_descriptors", s.Schema{
+					"min": c.Int("min"),
+					"max": c.Int("max"),
+					"avg": c.Int("avg"),
+				}),
+			}),
+			"jvm": c.Dict("jvm", s.Schema{
+				"max_uptime_in_millis": c.Int("max_uptime_in_millis"),
+				"mem": c.Dict("mem", s.Schema{
+					"heap_used_in_bytes": c.Int("heap_used_in_bytes"),
+					"heap_max_in_bytes":  c.Int("heap_max_in_bytes"),
+				}),
+				"threads": c.Int("threads"),
+			}),
+			"fs": c.Dict("fs", s.Schema{
+				"total_in_bytes":     c.Int("total_in_bytes"),
+				"free_in_bytes":      c.Int("free_in_bytes"),
+				"available_in_bytes": c.Int("available_in_bytes"),
+			}),
+		}),
+	}
+)
+
+func passthruField(fieldPath string, sourceData, targetData common.MapStr) error {
+	fieldValue, err := sourceData.GetValue(fieldPath)
+	if err != nil {
+		return elastic.MakeErrorForMissingField(fieldPath, elastic.Elasticsearch)
+	}
+
+	targetData.Put(fieldPath, fieldValue)
+	return nil
+}
+
+func clusterNeedsTLSEnabled(license, stackStats common.MapStr) (bool, error) {
+	// TLS does not need to be enabled if license type is something other than trial
+	value, err := license.GetValue("license.type")
+	if err != nil {
+		return false, err
+	}
+
+	licenseType, ok := value.(string)
+	if !ok {
+		return false, fmt.Errorf("License type is not a string")
+	}
+
+	if licenseType != "trial" {
+		return false, nil
+	}
+
+	// TLS does not need to be enabled if security is not enabled
+	value, err = stackStats.GetValue("security.enabled")
+	if err != nil {
+		return false, err
+	}
+
+	isSecurityEnabled, ok := value.(bool)
+	if !ok {
+		return false, fmt.Errorf("Security enabled flag is not a boolean")
+	}
+
+	if !isSecurityEnabled {
+		return false, nil
+	}
+
+	// TLS does not need to be enabled if TLS is already enabled on the transport protocol
+	value, err = stackStats.GetValue("security.ssl.transport.enabled")
+	if err != nil {
+		return false, err
+	}
+
+	isTLSAlreadyEnabled, ok := value.(bool)
+	if !ok {
+		return false, fmt.Errorf("Transport protocol SSL enabled flag is not a boolean")
+	}
+
+	return !isTLSAlreadyEnabled, nil
+}
+
+// computeNodesHash computes a simple hash value that can be used to determine if the nodes listing has changed since the last report.
+func computeNodesHash(clusterState common.MapStr) (int32, error) {
```

Comment:
@pickypg If possible I'd like you to review this function. It attempts to port over the logic implemented in https://github.com/elastic/elasticsearch/blob/93ada98eb54d99b9d98ed51778bd6422d18fd392/x-pack/plugin/monitoring/src/main/java/org/elasticsearch/xpack/monitoring/collector/cluster/ClusterStatsMonitoringDoc.java#L180-L195.